probe/consul.go

```diff
@@ -13,28 +13,73 @@ import (
 
 // S3Service describe a S3 service and associated metadata
 type S3Service struct {
-	Name    string
-	Gateway bool
+	Name                string
+	Endpoint            string
+	Gateway             bool
+	GatewayReadEnpoints []S3Endpoint
 }
 
-// NewProbeFromConsul Create a new probe using consul to generate endpoint configuration
-func NewProbeFromConsul(service S3Service, cfg config.Config, consulClient consul_api.Client, controlChan chan bool) (Probe, error) {
-	health := consulClient.Health()
-	serviceEntries, _, _ := health.Service(service.Name, "", false, nil)
+func (s *S3Service) Equals(other *S3Service) bool {
+	if s.Name != other.Name ||
+		s.Endpoint != other.Endpoint ||
+		s.Gateway != other.Gateway ||
+		len(s.GatewayReadEnpoints) != len(other.GatewayReadEnpoints) {
+		return false
+	}
 
-	endpoint, err := getEndpointFromConsul(service.Name, *cfg.EndpointSuffix, serviceEntries)
-	if err != nil {
-		log.Println("Could not generate endpoint from consul:", err)
-		return Probe{}, err
+	for i, gatewayReadEndPoint := range s.GatewayReadEnpoints {
+		if gatewayReadEndPoint.Name != other.GatewayReadEnpoints[i].Name {
+			return false
+		}
 	}
-	var readEndpoints []s3endpoint
-	if service.Gateway {
-		readEndpoints, err = extractGatewayEndoints(serviceEntries, cfg, consulClient)
-		if err != nil {
-			return Probe{}, err
+
+	return true
+}
+
+// GetS3Services Discover S3 services from consul (S3 service must match given consulTag or consulGatewayTag)
+func GetS3Services(cfg config.Config, consulClient consul_api.Client, consulTag string, consulGatewayTag string) []S3Service {
+	catalog := consulClient.Catalog()
+	health := consulClient.Health()
+	services, _, _ := catalog.Services(nil)
+
+	var s3Services []S3Service
+	var service S3Service
+	for serviceName := range services {
+		for i := range services[serviceName] {
+			isGateway := services[serviceName][i] == consulGatewayTag
+
+			if isGateway || services[serviceName][i] == consulTag {
+				serviceEntries, _, err := health.Service(service.Name, "", false, nil)
+				if err != nil {
+					log.Printf("Consul query failed for %s: %s", service.Name, err)
+					break
+				}
+
+				endpoint, err := getEndpointFromConsul(service.Name, *cfg.EndpointSuffix, serviceEntries)
+				if err != nil {
+					break
+				}
+
+				var readEndpoints []S3Endpoint
+				if isGateway {
+					readEndpoints, err = extractGatewayEndoints(serviceEntries, cfg, consulClient)
+					if err != nil {
+						break
```

Comment:
We should log the errors here and at L60. Maybe we can have a Prometheus metrics to track the discovery errors: probe_discovery_error_total (counter). This way we can add some monitoring if the probe is reporting to many errors

Reply:
done for logs

Reply:
done for metric